File: medium/447.number-of-boomerangs.java
import java.util.HashMap;

/*
 * @lc app=leetcode id=447 lang=java
 *
 * [447] Number of Boomerangs
 */

// @lc code=start
class Solution {
    public int numberOfBoomerangs(int[][] points) {
        HashMap<Integer, Integer> counterMap = new HashMap<>();
        int total = 0;

        for (int i=0; i<points.length; i++) {
            int x = points[i][0];
            int y = points[i][1];
            counterMap.clear();
            for (int j=0; j<points.length; j++) {
                int x1 = points[j][0];
                int y1 = points[j][1];

                int distance = (x-x1)*(x-x1) + (y-y1)*(y-y1);
                counterMap.put(distance, counterMap.getOrDefault(distance, 0) + 1);
            }

            for (int counter : counterMap.values()) {
                total += counter * (counter - 1);
            }
        }

        // [[0,0],[1,0],[-1,0],[0,1],[0,-1]]
        // [[1,1],[2,2],[3,3],[4,4],[5,5],[6,6],[7,7],[8,8]]
        return total;
    }
}
// @lc code=end

File: medium/447.number-of-boomerangs.js
/*
 * @lc app=leetcode id=447 lang=javascript
 *
 * [447] Number of Boomerangs
 * [hashmap]
 */

// @lc code=start
/**
 * @param {number[][]} points
 * @return {number}
 */
var numberOfBoomerangs = function(points) {
    // for each point i, calculate the dist of every other point j
    // use a hashmap to save, how many points have the same dist to point i
    // then, the boomerangs start with point i, is a permutation number of each dist
    // do a loop for every points and sum the numbers
    let ans = 0;
    for (let i=0; i<points.length; i++) {
      const distMap = new Map();
      for (let j=0; j<points.length; j++) {
        if (i == j) {
          continue;
        }
        const dist = Math.pow(points[i][0] - points[j][0], 2) +  Math.pow(points[i][1] - points[j][1], 2);
        const count = distMap.get(dist) || 0;
        distMap.set(dist, count + 1);
      }

      for (const entry of distMap) {
        ans += entry[1] * (entry[1] - 1);
      }
    }

    // [[0,0],[1,0],[-1,0],[0,1],[0,-1]]
    return ans;
};
// @lc code=end

